gateway/shard/manager.go

```diff
@@ -0,0 +1,375 @@
+//
+// If an error occurs, Close will attempt to close all remaining gateways
+// first, before returning. If multiple errors occur during that process, a
+// MultiError will be returned.
+func (m *Manager) Close() error {
+	return m.ApplyError(func(g *gateway.Gateway) error { return g.Close() }, true)
+}
+
+// Pause pauses all gateways managed by this Manager.
+//
+// If an error occurs, Pause will attempt to pause all remaining gateways
+// first, before returning. If multiple errors occur during that process, a
+// MultiError will be returned.
+func (m *Manager) Pause() error {
+	return m.ApplyError(func(g *gateway.Gateway) error { return g.Pause() }, true)
+}
+
+// UpdateStatus updates the status of all gateways handled by this Manager.
+//
+// If an error occurs, UpdateStatus will attempt to update the status of all
+// remaining gateways first, before returning. If multiple errors occur during
+// that process, a MultiError will be returned.
+func (m *Manager) UpdateStatus(d gateway.UpdateStatusData) error {
+	return m.ApplyError(func(g *gateway.Gateway) error { return g.UpdateStatus(d) }, true)
+}
+
+// RequestGuildMembers is used to request all members for a guild or a list of
+// guilds. When initially connecting, if you don't have the GUILD_PRESENCES
+// Gateway Intent, or if the guild is over 75k members, it will only send
+// members who are in voice, plus the member for you (the connecting user).
+// Otherwise, if a guild has over large_threshold members (value in the Gateway
+// Identify), it will only send members who are online, have a role, have a
+// nickname, or are in a voice channel, and if it has under large_threshold
+// members, it will send all members. If a client wishes to receive additional
+// members, they need to explicitly request them via this operation. The server
+// will send Guild Members Chunk events in response with up to 1000 members per
+// chunk until all members that match the request have been sent.
+//
+// Due to privacy and infrastructural concerns with this feature, there are
+// some limitations that apply:
+//
+// 	1. GUILD_PRESENCES intent is required to set presences = true. Otherwise,
+// 	   it will always be false
+// 	2. GUILD_MEMBERS intent is required to request the entire member
+// 	   list — (query=‘’, limit=0<=n)
+// 	3. You will be limited to requesting 1 guild_id per request
+// 	4. Requesting a prefix (query parameter) will return a maximum of 100
+// 	   members
+//
+// Requesting user_ids will continue to be limited to returning 100 members
+func (m *Manager) RequestGuildMembers(d gateway.RequestGuildMembersData) error {
+	return m.FromGuildID(d.GuildIDs[0]).RequestGuildMembers(d)
+}
+
+// onShardingRequired is the function stored as Gateway.OnShardingRequired
+// in every of the Manager's gateways.
+func (m *Manager) onShardingRequired() {
+	if m.rescaleExec.CompareAndSwap(false) {
+		m.Close()
+
+		if m.Rescale == nil {
+			return
+		}
+
+		m.mutex.Lock()
```

Comment:
This mutex should probably be moved to be acquired after `Rescale()`, just to prevent any potential deadlocks.